### ISSUE TITLE
a mistake in jupyter

### ISSUE THREAD

**Issue by CalebDu** [dlsyscourse/hw2#1]
<img width="1195" alt="image" src="https://user-images.githubusercontent.com/59528230/193222053-5d24017b-513f-4a09-8aee-304a7a2ba240.png">

bias should be initialized to zero.

**Comment by CalebDu** [comment]
because  HW2 forum  is not open for online students. I present some of what I think are wrong and ambiguous in hw2 code at github. 
1.  `dataloader ` class in question 4,  jupyter notebook  says `Combines a dataset and a sampler`. but there is no implementation of sampler in HW2.  So I just deal with this part by naive index. And i am not sure how to implement **shuffled** dataloader.  I do it by `np.random.shuffle` function in every epoch. However. there is no exact **assert** check for **shuffled dataloader** in test code.  I don't know whether my implementation of shuffled dataloader is right in next `mlp train` test.  different shuffled order may cause different final loss in "mlp train" test. 

2. `epoch` function in question 5,  the returne value includes `acc` and `loss`. jupyter have not detail about how to compute  `acc` metric in train mode. Parameters of model is updated in every batch. So the first batch may have more poor `acc` performace than last batch.   jupyter says `Returns the **average** accuracy (as a float) and the **average** loss over all samples (as a float).` i compute acc in 2 ways. One is mean value of acc per batch , the other is  sum of correct prediction in each batch divided by num of sample .  but there is a big difference betweent test code answer and 2 my solutions, and value of loss is same as test code answer.  

3. `epoch` function in question 5,  the following test code illustrate that a model without any train has 0.9164 acc in eval mode. it's so confusing. Mybe the acc is wrong?
<img width="946" alt="image" src="https://user-images.githubusercontent.com/59528230/193460185-21da4ecc-db02-4163-814d-ff7a1c586643.png">

I passed almost all the tests, except the `epoch` and` mlp train` test.
in conclution.  I know there will be many problems and room for improvement in the first release of dlsys class. i suggest that the every task in jupyter book could  contain more elaborated detail or hints. Because of answer in test code is fixed and some tasks with ambiguous description，I spent a lot of time going over and over the test code to figure out the true intention of these tasks. if there are more key hints or more detail of main procedure. students could save lots of time and avoid misunderstandings.

**Comment by CalebDu** [comment]
i encountered same problem in BatchNorm1d. Because i wrongly coding `w * (x - mean) / ((var + self.eps)**0.5) + b` to `(x - mean) / ((var + self.eps)**0.5)*w + b`. Maybe you get same mistake？ lol. i am not sure.

for sgd problem.  i also meet same problem. i solve it by browsing source code in pytorch. it cost me lots of time to find error.  referring to weight decay.  when you compute momentum item *u_i*,  you should consider gradient of L2 regularization.  you need to plus theta to theta.grad.  Then using the new grad to  compute *u_i*. the same goes for Adam

**Comment by weixliu** [comment]
Thank you. For the BatchNorm1d, I did not make the same mistake. But I just found the test precision was adjusted 4 days ago and I found the test passed with the adjusted precision. https://github.com/dlsyscourse/hw2/blob/80ed86bc630cc35441949fd74b2b23010788b698/tests/test_nn_and_optim.py#L771 . So I'm not sure the reason why you can pass the test case with original precision. 

Let me try your resolution of the SGD, it is a very method to refer to the source code of PyTorch. I also took a long time to debug the Homework. I think it's because the HW was becoming more and more complex.

**Comment by CalebDu** [comment]
> Thank you. For the BatchNorm1d, I did not make the same mistake. But I just found the test precision was adjusted 4 days ago and I found the test passed with the adjusted precision.
> 
> https://github.com/dlsyscourse/hw2/blob/80ed86bc630cc35441949fd74b2b23010788b698/tests/test_nn_and_optim.py#L771
> 
> . So I'm not sure the reason why you can pass the test case with original precision.
> Let me try your resolution of the SGD. I used the below equation to implement the sgd_momentum_weight_decay_1.
> 
> ```
> grad = param.grad.detach().cached_data
> u_(i+1) = momentum * u_i + (1 - momentum) * grad
> param = (1 - lr * weight_decay) * param - (lr * u_(i+1))
> ```
> 
> it is a very method to refer to the source code of PyTorch. I also took a long time to debug the Homework. I think it's because the HW was becoming more and more complex.

i pass all sgd test by following equations. these equations are different to counterparts in slices. So i think jupyter needs to provide  more hints.
```
grad = param.grad.detach().cached_data + weight_decay * param.detach()
u_(i+1) = momentum * u_i + (1 - momentum) * grad
param -= (lr * u_(i+1))
```

**Comment by weixliu** [comment]
> > Thank you. For the BatchNorm1d, I did not make the same mistake. But I just found the test precision was adjusted 4 days ago and I found the test passed with the adjusted precision.
> > https://github.com/dlsyscourse/hw2/blob/80ed86bc630cc35441949fd74b2b23010788b698/tests/test_nn_and_optim.py#L771
> > 
> > . So I'm not sure the reason why you can pass the test case with original precision.
> > Let me try your resolution of the SGD. I used the below equation to implement the sgd_momentum_weight_decay_1.
> > ```
> > grad = param.grad.detach().cached_data
> > u_(i+1) = momentum * u_i + (1 - momentum) * grad
> > param = (1 - lr * weight_decay) * param - (lr * u_(i+1))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > it is a very method to refer to the source code of PyTorch. I also took a long time to debug the Homework. I think it's because the HW was becoming more and more complex.
> 
> i pass all sgd test by following equations. these equations are different to counterparts in slices. So i think jupyter needs to provide more hints.
> 
> ```
> grad = param.grad.detach().cached_data + weight_decay * param.detach()
> u_(i+1) = momentum * u_i + (1 - momentum) * grad
> param -= (lr * u_(i+1))
> ```

Yes. Your equations could pass all SGD test cases.
What equation do you use in computing the BatchNorm1d? I just found that equation (9) is different from the other I searched in Pytorch. There was no m in there.
<img width="1121" alt="截屏2022-10-09 上午9 50 38" src="https://user-images.githubusercontent.com/4216446/194733833-03ef950a-1b69-4020-8f0a-cc9db128422f.png">

**Comment by Mmmofan** [comment]
> > > Thank you. For the BatchNorm1d, I did not make the same mistake. But I just found the test precision was adjusted 4 days ago and I found the test passed with the adjusted precision.
> > > https://github.com/dlsyscourse/hw2/blob/80ed86bc630cc35441949fd74b2b23010788b698/tests/test_nn_and_optim.py#L771
> > > 
> > > . So I'm not sure the reason why you can pass the test case with original precision.
> > > Let me try your resolution of the SGD. I used the below equation to implement the sgd_momentum_weight_decay_1.
> > > ```
> > > grad = param.grad.detach().cached_data
> > > u_(i+1) = momentum * u_i + (1 - momentum) * grad
> > > param = (1 - lr * weight_decay) * param - (lr * u_(i+1))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > it is a very method to refer to the source code of PyTorch. I also took a long time to debug the Homework. I think it's because the HW was becoming more and more complex.
> > 
> > 
> > i pass all sgd test by following equations. these equations are different to counterparts in slices. So i think jupyter needs to provide more hints.
> > ```
> > grad = param.grad.detach().cached_data + weight_decay * param.detach()
> > u_(i+1) = momentum * u_i + (1 - momentum) * grad
> > param -= (lr * u_(i+1))
> > ```
> 
> Yes. Your equations could pass all SGD test cases. What equation do you use in computing the BatchNorm1d? I just found that equation (9) is different from the other I searched in Pytorch. There was no m in there. <img alt="截屏2022-10-09 上午9 50 38" width="1121" src="https://user-images.githubusercontent.com/4216446/194733833-03ef950a-1b69-4020-8f0a-cc9db128422f.png">

I used equation `x = (x - mean) / (x + eps)**2 * weight + bias` in training mode and `x = (x - running_mean) / ((running_var + eps)**2)` in eval mode, which passed all tests of BatchNorm.

**Comment by weixliu** [comment]
> > Yes. Your equations could pass all SGD test cases. What equation do you use in computing the BatchNorm1d? I just found that equation (9) is different from the other I searched in Pytorch. There was no m in there. <img alt="截屏2022-10-09 上午9 50 38" width="1121" src="https://user-images.githubusercontent.com/4216446/194733833-03ef950a-1b69-4020-8f0a-cc9db128422f.png">
> 
> I used equation `x = (x - mean) / (std + eps)**0.5 * weight + bias` in training mode and `x = (x - running_mean) / ((running_var + eps)**0.5)` in eval mode, which passed all tests of BatchNorm.
> 
> But I met some problems with Adam optim, I passed all tests of SGD but can not pass adam_batchnorm_eval_mode test of Adam, did you make it?

Yes. I'm struggling with the Adam optim test cases. I passed all BatchNorm1D and LayerNorm1D test cases also all SGD test cases. But I still failed in `test_optim_adam_batchnorm_eval_mode_1` and `test_optim_adam_layernorm_1`. The test results are similar but the precision did not meet 1e-5. So currently, I still have no idea about these failures.
I suspect there might be some numerical issue in my implementation.

**Comment by Mmmofan** [comment]
> 

Haha I don't think it's caused by numerical precision, I met some problems seems like numerical precision while all because of error coding.

Well I failed `test_optim_adam_batchnorm_eval_mode_1` and `test_optim_adam_z_memory_check_1` (I passed SGD though), but passed all submisssion.

BTW, I passed all tests of `DropOut` but failed submission...

**Comment by weixliu** [comment]
> > 
> 
> Haha I don't think it's caused by numerical precision, I met some problems seems like numerical precision while all because of error coding.
> 
> Well I failed `test_optim_adam_batchnorm_eval_mode_1` and `test_optim_adam_z_memory_check_1` (I passed SGD though), but passed all submisssion.
> 
> BTW, I passed all tests of `DropOut` but failed submission...

test_optim_adam_z_memory_check_1 could be ignored because you might create less tensor. You could add grad into the computation graph to pass it, such as using grad instead of grad.data.

For DropOut, if it is an evaluation, you should not drop out the input.
I am also struggling failed test case with `test_mlp_resnet_forward_2`. Because `test_mlp_resnet_forward_1` has passed. So I suspect there might be some issue with my BatchNorm1D implementation.

Can you share your implementation BatchNorm1D and LayerNorm1D? I will share my implementaition below and delete them later.

**Comment by weixliu** [comment]
```
class LayerNorm1d(Module):
    def __init__(self, dim, eps=1e-5, device=None, dtype="float32"):
        super().__init__()
        self.dim = dim
        self.eps = eps
        ### BEGIN YOUR SOLUTION
        self.weight = Parameter(init.ones(dim, dtype=dtype))
        self.bias = Parameter(init.zeros(dim, dtype=dtype))\
        ### END YOUR SOLUTION

    #TODO ensure the dytpe should be float32
    def forward(self, x: Tensor) -> Tensor:
        ### BEGIN YOUR SOLUTION
        e_x = x.sum(axes=(1,)) / x.shape[1]
        b_e_x = e_x.reshape((x.shape[0], 1))
        b_e_x = b_e_x.broadcast_to(x.shape)
        v_x = ((x - b_e_x)**2).sum(axes=(1,)) / x.shape[1]
        denominator = (v_x + self.eps)**0.5
        denominator = denominator.reshape((x.shape[0], 1))
        denominator = denominator.broadcast_to(x.shape)
        numerator = x - e_x.reshape((x.shape[0], 1)).broadcast_to(x.shape)
        ele = numerator/denominator
        res = self.weight * ele + self.bias
        return res
        ### END YOUR SOLUTION
        
class BatchNorm1d(Module):
    def __init__(self, dim, eps=1e-5, momentum=0.1, device=None, dtype="float32"):
        super().__init__()
        self.dim = dim
        self.eps = eps
        self.momentum = momentum
        ### BEGIN YOUR SOLUTION
        self.weight = Parameter(init.ones(dim, dtype=dtype))
        self.bias = Parameter(init.zeros(dim, dtype=dtype))
        self.running_mean = init.zeros(dim, dtype=dtype)
        self.running_var = init.ones(dim, dtype=dtype)
        ### END YOUR SOLUTION

    #TODO ensure the dytpe should be float32
    def forward(self, x: Tensor) -> Tensor:
        ### BEGIN YOUR SOLUTION
        E_x = (x.sum(axes = (0,)) / x.shape[0])
        bc_E_x = E_x.reshape((1, x.shape[1]))
        bc_E_x = bc_E_x.broadcast_to(x.shape)
        temp = ops.power_scalar(x - bc_E_x, 2).sum(axes=(0,))
        V_x = temp / (x.shape[0])
        V_x_unbias = temp / (x.shape[0] - 1)
        
        self.running_mean = (1 - self.momentum) * self.running_mean + self.momentum * E_x
        self.running_var = (1 - self.momentum) * self.running_var + self.momentum * V_x_unbias
        t_E_x = E_x if self.training else self.running_mean
        t_V_x = V_x if self.training else self.running_var

        t_E_x = t_E_x.broadcast_to(x.shape)
        t_V_x = t_V_x.broadcast_to(x.shape)

        if not self.training:
          ele = (x - t_E_x) / (t_V_x + self.eps)**0.5
          return ele
        
        ele = (x - t_E_x) / (t_V_x + self.eps)**0.5
        bc_weight = self.weight.broadcast_to(ele.shape)
        bc_bias = self.bias.broadcast_to(ele.shape)
        return bc_weight * ele + bc_bias
        ### END YOUR SOLUTION

**Comment by Mmmofan** [comment]
> 

I passed DropOut now. Here's my implementation of BN and LN, I'll delete them 2 hours later:
```python
# batch norm:
def forward(self, x: Tensor) -> Tensor:
        ### BEGIN YOUR SOLUTION
        batch = x.shape[0]
        if self.training:
            mean = ops.summation(x, 0) / batch
            self.running_mean = self.momentum * mean + \
                (1 - self.momentum) * self.running_mean
            mean = ops.broadcast_to(ops.reshape(mean, (1, self.dim)), x.shape)

            std = ops.summation(ops.power_scalar(x - mean, 2), 0) / batch
            self.running_var = self.momentum * std + \
                (1 - self.momentum) * self.running_var
            std = ops.broadcast_to(ops.reshape(std, (1, self.dim)), x.shape)

            x = (x - mean) / ops.power_scalar(std + self.eps, 0.5) * \
                ops.broadcast_to(ops.reshape(self.weight, (1, self.dim)), x.shape) \
                + ops.broadcast_to(ops.reshape(self.bias, (1, self.dim)), x.shape)
            return x
        else:
            x = (x - self.running_mean) / ((self.running_var + self.eps) ** 0.5)
            return x
        ### END YOUR SOLUTION

# layer norm:
def forward(self, x: Tensor) -> Tensor:
        ### BEGIN YOUR SOLUTION
        batch = x.shape[0]
        mean = ops.broadcast_to(
            ops.reshape(ops.summation(x, 1) / self.dim, (batch, 1)),
            x.shape)
        std = ops.broadcast_to(ops.reshape(
            ops.summation(
                ops.power_scalar(x - mean, 2), 1
            ) / self.dim, (batch, 1)), x.shape)
        x = (x - mean) / ops.power_scalar(std + self.eps, 0.5) * \
            ops.broadcast_to(ops.reshape(self.weight, (1, self.dim)), x.shape) + \
            ops.broadcast_to(ops.reshape(self.bias, (1, self.dim)), x.shape)
        return x
        ### END YOUR SOLUTION
```

**Comment by weixliu** [comment]
You could delete them. I think our implementations are the same. Let me check my other codes, -_-.
Hard to debug, do you have any good ideas for debugging?

**Comment by Mmmofan** [comment]
> You could delete them. I think our implementations are the same. Let me check my other codes, -_-. Hard to debug, do you have any good ideas for debugging?

Just clone to local, and write another script, copy the test function code, set break point in IDE and you can debug it.

For example, create a 'debug.py' script, and copy function `learn_model_1d` to it (also with other utilities function which are used):
```python
# debug.py

def learn_model_1d(feature_size, nclasses, _model, optimizer, epochs=1, **kwargs):
    ...

out = learn_model_1d(64, 16, lambda z: nn.Sequential(nn.Linear(64, 32), nn.ReLU(), nn.Linear(32, 16)), ndl.optim.Adam, lr=0.001)
```

**Comment by lordidiot** [comment]
> For DropOut, if it is an evaluation, you should not drop out the input. I am also struggling failed test case with `test_mlp_resnet_forward_2`. Because `test_mlp_resnet_forward_1` has passed. So I suspect there might be some issue with my BatchNorm1D implementation.

@weixliu sorry to ping you on an old issue. I am also facing this problem now, do you remember what was the root-cause of the problem?
